services/secondary-storage/server/test/org/apache/cloudstack/storage/resource/TestAppender.java

```diff
@@ -0,0 +1,173 @@
+    }
+    protected void append(LoggingEvent loggingEvent) {
+        checkArgument(loggingEvent != null, "append requires a non-null loggingEvent");
+        final Level level = loggingEvent.getLevel();
+        checkState(expectedPatternResults.containsKey(level), "level " + level + " not supported by append");
+        for (final PatternResult patternResult : expectedPatternResults.get(level)) {
+            if (patternResult.getPattern().matcher(loggingEvent.getRenderedMessage()).matches()) {
+                patternResult.markFound();
+            }
+        }
+    }
+    public void close() {
+// Do nothing ...
+    }
+    public boolean requiresLayout() {
+        return false;
+    }
+    public void assertMessagesLogged() {
+        final List<String> unloggedPatterns = new ArrayList<>();
+        for (final Map.Entry<Level, Set<PatternResult>> expectedPatternResult : expectedPatternResults.entrySet()) {
+            for (final PatternResult patternResults : expectedPatternResult.getValue()) {
+                if (!patternResults.isFound()) {
+                    unloggedPatterns.add(format("%1$s was not logged for level %2$s",
+                            patternResults.getPattern().toString(), expectedPatternResult.getKey()));
+                }
+            }
+        }
+        if (!unloggedPatterns.isEmpty()) {
+            fail(Joiner.on(",").join(unloggedPatterns));
+        }
+    }
+    private static final class PatternResult {
+        private final Pattern pattern;
+        private boolean foundFlag = false;
+        private PatternResult(Pattern pattern) {
+            super();
+            this.pattern = pattern;
+        }
+        public Pattern getPattern() {
+            return pattern;
+        }
+        public void markFound() {
+        // This operation is thread-safe because the value will only ever be switched from false to true. Therefore,
+        // multiple threads mutating the value for a pattern will not corrupt the value ...
+            foundFlag = true;
+        }
+        public boolean isFound() {
+            return foundFlag;
+        }
+        @Override
+        public boolean equals(Object thatObject) {
+            if (this == thatObject) {
+                return true;
+            }
+            if (thatObject == null || getClass() != thatObject.getClass()) {
+                return false;
+            }
+            PatternResult thatPatternResult = (PatternResult) thatObject;
+            return foundFlag == thatPatternResult.foundFlag &&
+                    Objects.equal(pattern, thatPatternResult.pattern);
+        }
+        @Override
+        public int hashCode() {
+            return Objects.hashCode(pattern, foundFlag);
+        }
+        @Override
+        public String toString() {
+            return format("Pattern Result [ pattern: %1$s, markFound: %2$s ]", pattern.toString(), foundFlag);
+        }
+    }
+
+    public static final class TestAppenderBuilder {
```

Comment:
Minor Nit: I should have named this class ``Builder`` since it is already namespaced ``TestAppender``.   I would recommend that change since ``new TestAppender.Builder()`` is cleaner than ``newTestAppender.TestAppenderBuilder()``.  (Cosmetic I know, but it hits my naming OCD)